services/sharedchannel/sync_send.go

```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sharedchannel
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/mattermost/go-i18n/i18n"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/services/remotecluster"
+	"github.com/mattermost/mattermost-server/v5/utils"
+	"github.com/wiggin77/merror"
+)
+
+type syncTask struct {
+	id         string
+	channelId  string
+	remoteId   string
+	AddedAt    time.Time
+	retryCount int
+}
+
+func newSyncTask(channelId string, remoteId string) syncTask {
+	return syncTask{
+		id:        channelId + remoteId, // combination of channelId and remoteId to avoid duplicates
+		channelId: channelId,
+		remoteId:  remoteId,   // empty means update all remote clusters
+		AddedAt:   time.Now(), // entries can be sorted oldest to newest
+	}
+}
+
+// NotifyChannelChanged is called to indicate that a shared channel has been modified,
+// thus triggering an update to all remote clusters.
+func (scs *Service) NotifyChannelChanged(channelId string) {
+	task := newSyncTask(channelId, "")
+	scs.addTask(task)
+}
+
+func (scs *Service) addTask(task syncTask) {
+	scs.mux.Lock()
+	if _, ok := scs.tasks[task.id]; !ok {
+		scs.tasks[task.id] = task
+	}
+	scs.mux.Unlock()
+
+	// wake up the sync goroutine
+	select {
+	case scs.changeSignal <- struct{}{}:
+	default:
+		// that's ok, the sync routine is already busy
+	}
+}
+
+// syncLoop creates a pool of goroutines which wait for notifications of channel changes and
+// updates each remote based on those changes.
+func (scs *Service) syncLoop(done chan struct{}) {
+	// wait for channel changed signal and update for oldest channel id.
+	for {
+		select {
+		case <-scs.changeSignal:
+			for {
+				task, ok := scs.removeOldestTask()
+				if !ok {
+					break
+				}
+				if err := scs.processTask(task); err != nil {
+					scs.server.GetLogger().Error("Failed to update shared channel", mlog.String("channelId", task.channelId), mlog.String("remoteId", task.remoteId), mlog.Err(err))
+					// put task back into map so it will update again
+					scs.addTask(task)
+				}
+			}
+		case <-done:
+			return
+		}
+	}
+}
+
+// removeOldestTask removes and returns the oldest task in the task map.
+// If no tasks are available then false is returned.
+func (scs *Service) removeOldestTask() (syncTask, bool) {
+	scs.mux.Lock()
+	defer scs.mux.Unlock()
+
+	var oldestTask syncTask
+	var oldestKey string
+
+	for k, v := range scs.tasks {
+		if v.AddedAt.Before(oldestTask.AddedAt) || oldestTask.AddedAt.IsZero() {
+			oldestKey = k
+			oldestTask = v
+		}
```

Comment:
We're looking for the oldest task, not just a task older than the previous.  This will change anyway when I fix the thread issue described above.  The reason this list is walked here, is that I use a map to avoid dups and I want ordering.  I could avoid map, and check for dups when enquing but I'd rather let the thread pool do list walking than the caller, which is probably in the middle of handling a Rest API request.  Enqueue is fast, dequeue is slower.